src/java/org/apache/cassandra/service/reads/range/RangeCommandIterator.java

```diff
@@ -0,0 +1,291 @@
+        }
+    }
+
+    private void updateConcurrencyFactor()
+    {
+        liveReturned += counter.counted();
+
+        concurrencyFactor = computeConcurrencyFactor(totalRangeCount, rangesQueried, maxConcurrencyFactor, command.limits().count(), liveReturned);
+    }
+
+    @VisibleForTesting
+    static int computeConcurrencyFactor(int totalRangeCount, int rangesQueried, int maxConcurrencyFactor, int limit, int liveReturned)
+    {
+        maxConcurrencyFactor = Math.max(1, Math.min(maxConcurrencyFactor, totalRangeCount - rangesQueried));
+        if (liveReturned == 0)
+        {
+            // we haven't actually gotten any results, so query up to the limit if not results so far
+            Tracing.trace("Didn't get any response rows; new concurrent requests: {}", maxConcurrencyFactor);
+            return maxConcurrencyFactor;
+        }
+
+        // Otherwise, compute how many rows per range we got on average and pick a concurrency factor
+        // that should allow us to fetch all remaining rows with the next batch of (concurrent) queries.
+        int remainingRows = limit - liveReturned;
+        float rowsPerRange = (float) liveReturned / (float) rangesQueried;
+        int concurrencyFactor = Math.max(1, Math.min(maxConcurrencyFactor, Math.round(remainingRows / rowsPerRange)));
+        logger.trace("Didn't get enough response rows; actual rows per range: {}; remaining rows: {}, new concurrent requests: {}",
+                     rowsPerRange, remainingRows, concurrencyFactor);
+        return concurrencyFactor;
+    }
+
+    /**
+     * Queries the provided sub-range.
+     *
+     * @param replicaPlan the subRange to query.
+     * @param isFirst in the case where multiple queries are sent in parallel, whether that's the first query on
+     * that batch or not. The reason it matters is that whe paging queries, the command (more specifically the
+     * {@code DataLimits}) may have "state" information and that state may only be valid for the first query (in
+     * that it's the query that "continues" whatever we're previously queried).
+     */
+    private SingleRangeResponse query(ReplicaPlan.ForRangeRead replicaPlan, boolean isFirst)
+    {
+        PartitionRangeReadCommand rangeCommand = command.forSubRange(replicaPlan.range(), isFirst);
+        // If enabled, request repaired data tracking info from full replicas but
+        // only if there are multiple full replicas to compare results from
+        if (DatabaseDescriptor.getRepairedDataTrackingForRangeReadsEnabled()
+            && replicaPlan.contacts().filter(Replica::isFull).size() > 1)
+        {
+            command.trackRepairedStatus();
+            rangeCommand.trackRepairedStatus();
+        }
+
+        ReplicaPlan.SharedForRangeRead sharedReplicaPlan = ReplicaPlan.shared(replicaPlan);
+        ReadRepair<EndpointsForRange, ReplicaPlan.ForRangeRead> readRepair
+                = ReadRepair.create(command, sharedReplicaPlan, queryStartNanoTime);
```

Comment:
nit/formatting: We'd usually have `=` on the line above?

Reply:
Changed [here](https://github.com/apache/cassandra/pull/801/commits/1fa21b90296318c283d6edb7288e7523f7809f93), I was using [the original](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/service/StorageProxy.java#L2306-L2311) but less common style.